ddr/ddr/Models/GameEntities/Oncomer.swift

```diff
@@ -0,0 +1,103 @@
+
+import Foundation
+import SpriteKit
+
+class Oncomer: SKSpriteNode, Spawnable {
+  
+  let spawner: Spawner<Oncomer>
+  let emitters: [Emitter]
+  var collisionEffects: [Effect]
+  var passEffects: [Effect]
+  var goneEffects: [Effect]
+  
+  override var zPosition: CGFloat {
+    didSet {
+      // update sound
+      emitters.forEach({ $0.updateZ(zPosition) })
+      
+      // update visual
+      xScale = (GameScene.HORIZON - abs(zPosition)) / GameScene.HORIZON
+      yScale = xScale
+    }
+  }
+  
+  // override to update emitter(s)
+  override var position : CGPoint {
+    didSet {
+      emitters.forEach({ $0.updatePosition(position) })
+    }
+  }
+  
+  init (
+    spawner: Spawner<Oncomer>,
+    emitters: [Emitter],
+    collisionEffects: [Effect],
+    passEffects: [Effect],
+    goneEffects: [Effect],
+    texture: SKTexture?,
+    color: UIColor,
+    size: CGSize)
+  {
+    self.spawner = spawner
+    self.collisionEffects = collisionEffects
+    self.passEffects = passEffects
+    self.goneEffects = goneEffects
+    self.emitters = emitters
+    super.init(texture: texture, color: color, size: size)
+    self.collisionEffects = self.collisionEffects + [ KillOncomerEffect(oncomer: self) ]
+    self.passEffects = self.passEffects + [ HideEffect(nodeToHide: self) ]
+    self.goneEffects = self.goneEffects + [ KillOncomerEffect(oncomer: self) ]
+  }
+  
+  required init?(coder aDecoder: NSCoder) {
+    fatalError("init(coder:) has not been implemented")
+  }
+  
+  private func getImage() -> String {
+    return ""
+  }
+  
+  func despawn() {
+    spawner.despawn(self)
+  }
+  
+  func collidesWith(position: CGPoint) -> Bool {
+    return false
+  }
+  
+//  could not use because rider headPosition != rider position
+//  func collidesWith(node: SKNode) -> Bool {
+//    return collidesWith(position: node.position)
+//  }
+  
+  func move() {
+    self.zPosition = self.zPosition - self.speed
+  }
+  
+  func move(withAdditionalDistance distance : CGFloat) {
+    self.zPosition = self.zPosition - self.speed - distance
+  }
+  
+  func isGone() -> Bool {
```

Comment:
So this will return true when the zPos is like, -101? Does that mean sounds will continue behind the player?

Reply:
It does, and I like it.